File: src/jkorn/pvpcore/world/areas/PvPCArea.php
<?php
/**
 * Created by PhpStorm.
 * User: jkorn2324
 * Date: 2019-05-19
 * Time: 20:18
 */

declare(strict_types=1);

namespace jkorn\pvpcore\world\areas;


use jkorn\pvpcore\utils\IKBObject;
use jkorn\pvpcore\utils\PvPCKnockback;
use jkorn\pvpcore\utils\Utils;
use pocketmine\level\Level;
use pocketmine\math\Vector3;
use pocketmine\Player;
use pocketmine\Server;


class PvPCArea implements IKBObject
{

    /** @var Vector3 */
    private $firstPos, $secondPos;
    /** @var PvPCKnockback */
    private $knockback;

    /** @var string */
    private $name;

    /** @var bool */
    private $enabled;

    /** @var Level|null */
    private $level;

    /**
     * PvPCArea constructor.
     * @param string $name
     * @param string|Level|null $level
     * @param Vector3 $firstPos
     * @param Vector3 $secondPos
     * @param PvPCKnockback $knockback
     * @param bool $enabled
     */
    public function __construct(string $name, $level, Vector3 $firstPos, Vector3 $secondPos, PvPCKnockback $knockback, bool $enabled = true)
    {
        $this->name = $name;
        $this->firstPos = $firstPos;
        $this->secondPos = $secondPos;
        $this->level = $level instanceof Level ? $level : is_string($level) ? Server::getInstance()->getLevelByName($level) : null;
        $this->knockback = $knockback;
        $this->enabled = $enabled;

    }

    /**
     * @param Vector3 $pos
     * @param bool $pos2
     *
     * Updates the position.
     */
    public function setPosition(Vector3 $pos, bool $pos2): void
    {

        $position = &($pos2 ? $this->secondPos : $this->firstPos);
        $position = $pos;
    }

    /**
     * @param bool $result
     *
     * Sets the enable.
     */
    public function setKBEnabled(bool $result): void
    {
        $this->enabled = $result;
    }

    /**
     * @return bool
     *
     * Determines if the kb is enabled or not.
     */
    public function isKBEnabled(): bool
    {
        return $this->enabled;
    }

    /**
     * @return PvPCKnockback
     *
     * Gets the knockback values for the area.
     */
    public function getKnockback(): PvPCKnockback
    {
        return $this->knockback;
    }

    /**
     * @return string
     *
     * Gets the name of the Area.
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @return Level|null
     */
    public function getLevel(): ?Level
    {
        return $this->level;
    }

    /**
     * @param Vector3 $pos
     * @return bool
     *
     * Determines if the position is within the area.
     */
    private function isWithinArea(Vector3 $pos) : bool
    {
        $maxX = Utils::getMaxValue((float)$this->firstPos->x, (float)$this->secondPos->x);
        $minX = Utils::getMinValue((float)$this->firstPos->x, (float)$this->secondPos->x);
        $maxY = Utils::getMaxValue((float)$this->firstPos->y, (float)$this->secondPos->y);
        $minY = Utils::getMinValue((float)$this->firstPos->y, (float)$this->secondPos->y);
        $maxZ = Utils::getMaxValue((float)$this->firstPos->z, (float)$this->secondPos->z);
        $minZ = Utils::getMinValue((float)$this->firstPos->z, (float)$this->secondPos->z);

        return $pos->x >= $minX && $pos->x <= $maxX
            && $pos->y >= $minY && $pos->y <= $maxY
            && $pos->z >= $minZ && $pos->z <= $maxZ;
    }

    /**
     * @return array
     *
     * Converts the area to an array.
     */
    public function toArray() : array
    {
        return [
            "enabled" => $this->enabled,
            "first-pos" => Utils::vec3ToArr($this->firstPos),
            "second-pos" => Utils::vec3ToArr($this->secondPos),
            "knockback" => $this->knockback->toArray(),
            "level" => $this->level !== null ? $this->level->getName() : null
        ];
    }

    /**
     * @param $object
     * @return bool
     *
     * Determines if the pvp area is equal to another.
     */
    public function equals($object) : bool
    {

        if ($object instanceof PvPCArea)
        {
            return $object->getName() === $this->name
                && $object->isKBEnabled() === $this->enabled;
        }

        return false;
    }


    /**
     * @param Player $player1
     * @param Player $player2
     * @return bool
     *
     * Determines if the player can use the custom knockback.
     */
    public function canUseKnocback(Player $player1, Player $player2): bool
    {
        if(!$this->level instanceof Level || !$this->enabled)
        {
            return false;
        }

        $level = $player1->getLevel();
        if(!Utils::areLevelsEqual($level, $player2->getLevel())
        || !Utils::areLevelsEqual($level, $this->level))
        {
            return false;
        }

        return $this->isWithinArea($player1->asVector3())
            && $this->isWithinArea($player2->asVector3());
    }

    /**
     * @param string $name
     * @param $data
     * @return PvPCArea|null
     *
     * Decodes the area based on the name and the data.
     */
    public static function decode(string $name, $data): ?PvPCArea
    {
        if(isset($data["enabled"], $data["first-pos"], $data["second-pos"], $data["knockback"], $data["level"]))
        {
            $server = Server::getInstance();

            $level = $data["level"];
            if($level !== null)
            {
                $server->loadLevel($level);
            }

            $enabled = (bool)$data["enabled"];
            $firstPos = Utils::arrToVec3($data["first-pos"]);
            $secondPos = Utils::arrToVec3($data["second-pos"]);
            $knockback = PvPCKnockback::decode($data["knockback"]);

            if($firstPos !== null && $secondPos !== null && $knockback !== null)
            {
                return new PvPCArea(
                    $name,
                    $level,
                    $firstPos,
                    $secondPos,
                    $knockback,
                    $enabled
                );
            }
        }

        return null;
    }

    /**
     * @param string $name
     * @param $data
     * @return PvPCArea|null
     *
     * Decodes the PvPCArea based on the data from the old information.
     */
    public static function decodeLegacy(string $name, $data): ?PvPCArea
    {
        if(isset($data["level"], $data["kb"], $data["attack-delay"], $data["enabled"], $data["first-pos"], $data["second-pos"]))
        {
            $server = Server::getInstance();

            $level = $data["level"];
            if($level !== null)
            {
                $server->loadLevel($level);
            }

            $enabled = (bool)$data["enabled"];
            $firstPos = Utils::arrToVec3($data["first-pos"]);
            $secondPos = Utils::arrToVec3($data["second-pos"]);

            if($firstPos !== null && $secondPos !== null)
            {
                return new PvPCArea(
                    $name,
                    $level,
                    $firstPos,
                    $secondPos,
                    new PvPCKnockback((float)$data["kb"], (float)$data["kb"], (int)$data["attack-delay"]),
                    $enabled
                );
            }
        }

        return null;
    }
}

File: Nukkit/src/pvpcore/worlds/areas/PvPCArea.java
package pvpcore.worlds.areas;

import cn.nukkit.Player;
import cn.nukkit.Server;
import cn.nukkit.level.Level;
import cn.nukkit.math.Vector3;
import pvpcore.utils.IKnockbackObject;
import pvpcore.utils.PvPCKnockback;
import pvpcore.utils.Utils;

import java.util.HashMap;
import java.util.Map;

public class PvPCArea implements IKnockbackObject {

    /** The level that the pvp area is in.*/
    private Level level;
    /** The knockback variable. */
    private PvPCKnockback knockback;
    /** The position boundaries of the PvPArea. */
    private Vector3 firstPosition, secondPosition;
    /** The name of the area. */
    private String name;
    /** Determines if the PvPArea KB is enabled. */
    private boolean enabled;

    /**
     * The constructor for the PvPArea.
     * @param name - The name of the area.
     * @param level - The level of the area.
     * @param firstPosition - The first position of the area.
     * @param secondPosition - The second position of the area.
     */
    public PvPCArea(String name, Level level, Vector3 firstPosition, Vector3 secondPosition)
    {
        this.name = name;
        this.level = level;
        this.firstPosition = firstPosition;
        this.secondPosition = secondPosition;
        this.knockback = new PvPCKnockback();
        this.enabled = true;
    }

    /**
     * The constructor for the PvPArea.
     * @param name - The name of the PvPArea.
     * @param level - The level of the PvPArea.
     * @param firstPosition - The first position of the area.
     * @param secondPosition - The second position of area.
     * @param knockback - The knockback input of the PvPArea.
     * @param enabled - Determines whether or not the PvPArea is enabled.
     */
    public PvPCArea(String name, String level, Vector3 firstPosition, Vector3 secondPosition, PvPCKnockback knockback, boolean enabled)
    {
        this.name = name;
        this.level = Server.getInstance().getLevelByName(level);
        this.firstPosition = firstPosition;
        this.secondPosition = secondPosition;
        this.knockback = knockback;
        this.enabled = enabled;
    }


    /**
     * Sets the corresponding position.
     * @param position - The Vector3 position variable.
     * @param pos2 - Determines if we are setting the position to the second one.
     */
    public void setPosition(Vector3 position, boolean pos2)
    {
        if(pos2) {
            this.secondPosition = position;
        } else {
            this.firstPosition = position;
        }
    }

    /**
     * Sets the area knockback as enabled or not.
     * @param enabled - Determines whether kb is enabled.
     */
    public void setEnabled(boolean enabled)
    {
        this.enabled = enabled;
    }

    /**
     * Determines whether the KB is enabled.
     * @return - True if enabled, false otherwise.
     */
    public boolean isEnabled()
    {
        return this.enabled;
    }

    /**
     * Gets the knockback of the area.
     * @return - The knockback object.
     */
    public PvPCKnockback getKnockback()
    {
        return this.knockback;
    }

    /**
     * Gets the name of the area.
     * @return - The name of the area.
     */
    public String getName()
    {
        return this.name;
    }

    /**
     * Gets the level of the area.
     * @return - The level or null.
     */
    public Level getLevel()
    {
        return this.level;
    }

    /**
     * Determines if each player can use the knockback.
     * @param player - The player object.
     * @param attacker - The player who is attacking the player.
     * @return true if the players can use the knockback.
     */
    @Override
    public boolean canUseKnockback(Player player, Player attacker)
    {
        if(this.level == null || !this.enabled)
        {
            return false;
        }

        Level player1Level = player.getLevel();
        if(
                !Utils.levelsEqual(player1Level, attacker.getLevel())
                || !Utils.levelsEqual(player1Level, this.level)
        )
        {
            return false;
        }

        return this.isWithinArea(player) && this.isWithinArea(attacker);
    }

    /**
     * Determines whether the position is within the provided area.
     * @param position - The input position.
     * @return - True if the position is within the area, false otherwise.
     */
    private boolean isWithinArea(Vector3 position)
    {
        double minX = (double) Utils.getMinValue(this.firstPosition.x, this.secondPosition.x);
        double maxX = (double) Utils.getMaxValue(this.firstPosition.x, this.secondPosition.x);
        double minY = (double) Utils.getMinValue(this.firstPosition.y, this.secondPosition.y);
        double maxY = (double)Utils.getMaxValue(this.firstPosition.y, this.secondPosition.y);
        double minZ = (double)Utils.getMinValue(this.firstPosition.z, this.secondPosition.z);
        double maxZ = (double) Utils.getMaxValue(this.firstPosition.z, this.secondPosition.z);

        return position.x >= minX && position.x <= maxX
                && position.y >= minY && position.y <= maxY
                && position.z >= minZ && position.z <= maxZ;
    }

    /**
     * Exports the value to a HashMap.
     * @return The hashmap containing the data.
     */
    @Override
    public HashMap<String, Object> export()
    {
        HashMap<String, Object> output = new HashMap<>();
        output.put("enabled", this.enabled);
        output.put("first-pos", Utils.vec3ToMap(this.firstPosition));
        output.put("second-pos", Utils.vec3ToMap(this.secondPosition));
        output.put("knockback", this.knockback.export());
        output.put("level", this.level != null ? this.level.getName() : null);
        return output;
    }

    /**
     * Determines if an object is equivalent to another object.
     * @param object - The input object.
     * @return - True if the area is equivalent, false otherwise.
     */
    public boolean equals(Object object)
    {
        if(object instanceof PvPCArea)
        {
            return ((PvPCArea) object).name.equals(this.name);
        }

        return false;
    }

    /**
     * Decodes the data into a new PvPCArea instance.
     * @param name - The name of the PvPArea.
     * @param data - The data value of the PvPArea.
     * @return - Null if unsuccessful, PvPCArea instance otherwise.
     */
    public static PvPCArea decode(String name, Object data)
    {
        if(
                data instanceof Map
                && ((Map) data).containsKey("enabled")
                && ((Map) data).containsKey("first-pos")
                && ((Map) data).containsKey("second-pos")
                && ((Map) data).containsKey("knockback")
                && ((Map) data).containsKey("level")
        )
        {
            String levelName = String.valueOf(((Map) data).get("level"));
            if(levelName != null && !Server.getInstance().isLevelLoaded(levelName))
            {
                Server.getInstance().loadLevel(levelName);
            }

            Object enabled = ((Map) data).get("enabled");
            Vector3 firstPos = Utils.mapToVec3(((Map) data).get("first-pos"));
            Vector3 secondPos = Utils.mapToVec3(((Map) data).get("second-pos"));
            PvPCKnockback knockback = PvPCKnockback.decode(((Map) data).get("knockback"));

            if(
                    enabled instanceof Boolean
                    && firstPos != null
                    && secondPos != null
                    && knockback != null
            )
            {
                return new PvPCArea(
                        name,
                        levelName,
                        firstPos,
                        secondPos,
                        knockback,
                        (Boolean) enabled
                );
            }
        }

        return null;
    }

    /**
     * Decodes the data into a new PvPCArea instance based on legacy format.
     * @param name - The name of the PvPArea.
     * @param data - The data value of the PvPArea.
     * @return - Null if unsuccessful, PvPCArea instance otherwise.
     */
    public static PvPCArea decodeLegacy(String name, Object data)
    {
        if(
                data instanceof Map
                && ((Map) data).containsKey("knockback-y")
                && ((Map) data).containsKey("knockback-xz")
                && ((Map) data).containsKey("first-pos")
                && ((Map) data).containsKey("second-pos")
                && ((Map) data).containsKey("attack-delay")
                && ((Map) data).containsKey("kb-enabled")
                && ((Map) data).containsKey("world")
        )
        {
            String levelName = (String)((Map) data).get("world");
            if(levelName != null && !Server.getInstance().isLevelLoaded(levelName))
            {
                Server.getInstance().loadLevel(levelName);
            }

            Number kbY = (Number)((Map) data).get("knockback-y");
            Number kbXZ = (Number)((Map) data).get("knockback-xz");
            Vector3 firstPos = stringToVec3(String.valueOf(((Map) data).get("first-pos")));
            Vector3 secondPos = stringToVec3(String.valueOf(((Map) data).get("second-pos")));
            Number attackDelay = (Number) ((Map) data).get("attack-delay");
            boolean enabled = Boolean.parseBoolean(String.valueOf(((Map) data).get("kb-enabled")));

            if(
                    firstPos != null
                    && secondPos != null
            )
            {
                return new PvPCArea(
                        name,
                        levelName,
                        firstPos,
                        secondPos,
                        new PvPCKnockback(kbY.floatValue(), kbXZ.floatValue(), attackDelay.intValue()),
                        enabled
                );
            }
        }
        return null;
    }

    /**
     * Converts the input string to a vector 3.
     * @param input - The input string.
     * @return Vector3 or null.
     */
    private static Vector3 stringToVec3(String input)
    {
        String[] exploded = input.split(":");
        if(exploded.length == 3)
        {
            int x = Integer.parseInt(exploded[0]);
            int y = Integer.parseInt(exploded[1]);
            int z = Integer.parseInt(exploded[2]);
            return new Vector3(x, y, z);
        }
        return null;
    }
}
